File: web/data/entities/route.py
class Route:
    def __init__(self, id, unique_id, stops, day, company, info = ""):
        self.id = id
        self.unique_id = unique_id
        self.stops = stops
        self.day = day
        self.company = company
        self.info = info

    def getStopIdx(self, stop, position):
        return stops[stop]

    def __str__(self):
        return 


'''
class Route constructor(val id: String, val unique_id: String, val stops: Map<Stop, List<String>>, val day: TypeOfDay, val company: Int, val info: String? = "") {
    private var allStops: List<Stop> = stops.keys.toList()
    private var origin: Stop? = allStops[0]
    private var destination: Stop? = allStops[allStops.size - 1]

    init {
        if (Datasource().getRouteHash().containsKey(unique_id)) Log.w("ERROR", "$unique_id duplicated")
        else Datasource().addRouteToHash(unique_id, this)
    }

    fun getStopTime(stop: Stop?, position: Int): String? {
        return stops[stop]?.get(position)
    }

    fun getTimeIdx(times: List<String>, time: String?): Int {
        for (i in times.indices)
            if (times[i] == time) return i
        Log.d("dl", times.toString() + time)
        return -1
    }

    fun getStopIdx(stop: Stop?): Int {
        var i : Int = 0
        for (st in stops.keys)
            if (st == stop) return i
            else i++
        return -1
    }

    fun getOrigin(): Stop? {
        return origin
    }

    fun getDestination(): Stop? {
        return destination
    }

    fun getNStops(stop: Stop?): Int? {
        return stops[stop]?.size
    }
}
'''

File: src/saomiguelbus/lib/models/route.dart
import 'index.dart';

class Route {
  final String id;
  final String uniqueId;
  final Map<Stop, String> stops;
  final TypeOfDay day;
  final Company company;
  final String? info;

  List<Stop> allStops = [];
  Stop? origin;
  Stop? destination;

  Route(this.id, this.uniqueId, this.stops, this.day, this.company,
      {this.info}) {
    allStops = stops.keys.toList();
    origin = allStops[0];
    destination = allStops[allStops.length - 1];

    // if (Datasource().getRouteHash().containsKey(uniqueId)) {
    //   developer.log('ERROR: $uniqueId duplicated');
    // } else {
    //   Datasource().addRouteToHash(uniqueId, this);
    // }
  }

  //TODO: Refactor this bcoz it's not a list anymore
  (Stop?, String?) getStopTime(String stopName) {
    for (Stop stop in stops.keys) {
      if (stop.name == stopName) {
        return (stop, stops[stop]);
      }
    }
    return (null, null);
  }

  //TODO: Refactor this bcoz it's not a list anymore
  int getTimeIdx(List<String> times, String? time) {
    for (int i = 0; i < times.length; i++) {
      if (times[i] == time) {
        return i;
      }
    }
    return -1;
  }

  //TODO: Refactor this bcoz it's not a list anymore
  int getStopIdx(Stop? stop) {
    int i = 0;
    for (Stop st in stops.keys) {
      if (st == stop) {
        return i;
      } else {
        i++;
      }
    }
    return -1;
  }

  Stop? getOrigin() {
    return origin;
  }

  Stop? getDestination() {
    return destination;
  }

  int? getNStops(Stop? stop) {
    return stops[stop]?.length;
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Route &&
          runtimeType == other.runtimeType &&
          uniqueId == other.uniqueId;

  @override
  int get hashCode => uniqueId.hashCode;

  @override
  String toString() {
    return '$id | $stops';
  }
}
